File: src/euler8.rs
use std::io::{BufRead, BufReader};
use std::fs::File;
use std::cmp::max;

const ASCII_NEWLINE: u8 = 0xa;

fn product ( digits: &[u8] ) -> u64 {
	digits.iter().fold(1, |acc, &x| acc * (x as u64))
}

fn load_digits( filepath: & str ) -> Vec<u8> {
	let f = File::open(filepath).unwrap();
	let mut reader = BufReader::new(f);
	let mut output = Vec::new();
	reader.read_until(ASCII_NEWLINE, &mut output).unwrap();

	if output[output.len() - 1] == ASCII_NEWLINE {
		output.pop();
	}

	// assume that the file is encoded in something ascii compatible.
	// Then the character '0' is encoded as 0x30, so we should subtract this to get the actual number.
	for index in 0..output.len() {
		output[index] -= 0x30;
	}
	output
}

fn max_product_of_substring(digits: Vec<u8>, width: usize) -> u64 {
	if width > digits.len() {
		panic!("Not enough digits to find a product of length {}", width);
	};
	let mut max_product = 0;
	for index in 0..digits.len()-width {
		max_product = max(max_product, product(&digits[index..index+width]));
	}
	max_product
}

/// Find the thirteen adjacent digits in the 1000-digit number that have the greatest product. What is the value of this product?
fn main() {
	let digits = load_digits("resource/problem8.num");
	println!("{}", max_product_of_substring(digits, 13) );
}

/// The four adjacent digits in the 1000-digit number that have the greatest product are 9 × 9 × 8 × 9 = 5832.
#[test]
fn test() {
	let expected = 5832;
	let digits = load_digits("resource/problem8.num");
	let result = max_product_of_substring(digits, 4);
	assert_eq!(result, expected);
}

#[test]
fn test_product() {
	let expected = 24;
	let digits = vec![1, 2, 3, 4];
	let result = product(&digits);
	assert_eq!(result, expected);
}

File: c/src/euler8.c
#include <stdio.h>
#include <assert.h>

#define NUM_DIGITS 1000

/*
The four adjacent digits in the 1000-digit number that have the greatest product are 9 × 9 × 8 × 9 = 5832.
Find the thirteen adjacent digits in the 1000-digit number that have the greatest product. What is the value of this product?
*/
long unsigned max(long unsigned a, long unsigned b){
	return a < b ? b : a;
}

long unsigned sweep(unsigned* array, size_t index, size_t num_factors){
	long unsigned prod = 1;
	for (size_t j = 0; j < num_factors; j++){
		prod *= array[index + j];
	}
	return prod;
}

long unsigned max_moving_product(unsigned* array, size_t array_length, size_t num_factors){
	long unsigned max_prod = 0;
	size_t limit = array_length - num_factors;
	for (size_t i = 0; i <= limit; i++){
		max_prod = max(max_prod, sweep(array, i, num_factors));
	}
	return max_prod;
}

void setup(unsigned* data, char const* filename){
	char buffer [NUM_DIGITS];
	FILE* source = fopen(filename, "r");
	fread(buffer, sizeof(char), NUM_DIGITS, source);
	fclose(source);
	for (int i = 0; i < NUM_DIGITS; i += 1){
			data[i] = buffer[i] - '0';
	}
}

int main(){
	unsigned data[NUM_DIGITS];
	setup(data, "../resource/problem8.num");
	assert(max_moving_product(data, NUM_DIGITS, 4) == 5832);
	printf("%lu\n", max_moving_product(data, NUM_DIGITS, 13));
	return 0;
}